torch/export/_safeguard.py

```diff
@@ -0,0 +1,15 @@
+import torch
+from torch.overrides import TorchFunctionMode
+
+
+class GradModeUnsupportedSafeguard(TorchFunctionMode):
+    def __torch_function__(self, func, types, args=(), kwargs=None):
+        kwargs = kwargs or {}
+        unsupported_grad_mode_ops = [
```

Comment:
If you are going to add the other similar ops in the follow up PR, can you leave a TODO here?

Reply:
I regard the other ops to be added as a new task instead of a follow-up. For now, I don't have ideas about them. You may elaborate later or now what the TODO is so that I can understand what's the relation for this one and future PRs.

Reply:
We right now don't handle:

- Custom autograd functions
- x.requires_grad_
- x.requires_grad = foo: bool

At least the latter two might be doable within this PR, can you add test cases for the latter two and see if we can intercept it with our new mode?

Reply:
For custom autograd functions, dynamo should make them as special operators, so you can write a test case and see how dynamo handles them. We could probably just error by checking the dynamo produced graph has such special ops. This part can be done as a seperate PR

Reply:
Current dynamo itself has stopped `Tensor.requires_grad_` (https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/tensor.py#L756C1-L759C51) and `Tensor.requires_grad = ` (https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/builtin.py#L1289C1-L1294C22). So we don't need to worry them for now. 

Do you think we should add a test to guarantee they are unsupported from export API side?

Reply:
What's about **Custom autograd functions**?

Reply:
We should error when we see a custom autograd function.

Reply:
Can you give an example? What's the definition of autograd function?